### PR TITLE
std: refactor `pthread`-based synchronization

### PR DESCRIPTION
**Pull request by joboet**
The non-trivial code for `pthread_condvar` is duplicated across the thread parking and the `Mutex`/`Condvar` implementations. This PR moves that code into `sys::pal`, which now exposes an `unsafe` wrapper type for `pthread_mutex_t` and `pthread_condvar_t`.

### CODE REVIEW

**Comment by ChrisDenton** [comment]
This looks good to me on the surface but I don't feel I can confidently r+ atm as I don't currently have the time to go over it properly. So I'll reroll.

r? libs

**Review by workingjubilee — CHANGES_REQUESTED** [review]
So, it seems like this is going to be a big diff either way. But is there any chance you can do the "coalesce in a single spot" and "change up various internals" in separate PRs? Or at least commits?

In particular, if possible, I'd like for the LazyBox -> OnceBox to be separately bisectable, even if it means I review somewhat gnarlier code.

**Comment by joboet** [comment]
I've merged the `OnceBox` change in #131094, so this just contains the deduplication and wrapper creation.

@rustbot ready

File: library/std/src/sys/sync/condvar/pthread.rs [hunk 1]
```diff
@@ -1,196 +1,88 @@
...
-    fn get(&self) -> *mut libc::pthread_cond_t {
-        self.inner.get_or_init(AllocatedCondvar::new).0.get()
+    #[inline]
+    fn get(&self) -> Pin<&pal::Condvar> {
+        self.cvar.get_or_init(|| {
+            let mut cvar = Box::pin(pal::Condvar::new());
+            // SAFETY: we only call `init` once, namely here.
+            unsafe { cvar.as_mut().init() };
+            cvar
+        })
```
Comment (workingjubilee) [inline library/std/src/sys/sync/condvar/pthread.rs:28]: So, as I understand it, the closure can potentially get called twice, as the sync can race. That means that `init` *has to be safe to call twice*, even if it means that it can be "safe, but only if you immediately destroy it if it's a copy". That means that this `// SAFETY` comment is somewhat off-base.

File: library/std/src/sys/sync/condvar/pthread.rs [hunk 2]
```diff
@@ -1,196 +1,88 @@
...
+        // SAFETY: the caller guarantees that the lock is owned, thus the mutex
+        // must have been initialized already.
+        let mutex = unsafe { mutex.pal.get_unchecked() };
         self.verify(mutex);
-        let r = libc::pthread_cond_wait(self.get(), mutex);
-        debug_assert_eq!(r, 0);
+        // SAFETY: we called `init` above, we verified that this condition
+        // variable is only used with `mutex` and the caller guarantees that
+        // `mutex` is locked by the current thread.
+        unsafe { self.get().wait(mutex) }
```
Comment (workingjubilee) [inline library/std/src/sys/sync/condvar/pthread.rs:64]: A lot of stuff is riding on this `init` being correct, too...

File: library/std/src/sys/sync/mutex/pthread.rs [hunk 3]
```diff
@@ -1,163 +1,66 @@
...
-        self.inner.get_or_init(AllocatedMutex::new).0.get()
+    fn get(&self) -> Pin<&pal::Mutex> {
+        // If the initialization race is lost, the new mutex is destroyed.
+        // This is sound however, as it cannot have been locked.
+        self.pal.get_or_init(|| {
+            let mut pal = Box::pin(pal::Mutex::new());
+            // SAFETY: we only call `init` once, namely here.
+            unsafe { pal.as_mut().init() };
+            pal
+        })
```
Comment (workingjubilee) [inline library/std/src/sys/sync/mutex/pthread.rs:27]: I believe this is the complete and correct comment?

**Review by workingjubilee — CHANGES_REQUESTED** [review]
Sorry about taking so long, it was kinda hard to follow all the moves.

We need to get more pedantic.

**Comment by workingjubilee** [comment]
Passed on these jobs:
- try-job: dist-various-1
- try-job: dist-various-2
- try-job: test-various
- try-job: armhf-gnu
- try-job: aarch64-gnu
- try-job: x86_64-fuchsia
- try-job: dist-i686-linux
- try-job: aarch64-apple
- try-job: dist-android
- try-job: dist-x86_64-freebsd

File: library/std/src/sys/sync/condvar/pthread.rs (hunk 1 above)
Comment (joboet) [reply library/std/src/sys/sync/condvar/pthread.rs:28]: The invariant is that `init` may only be called once per object; that isn't violated by this.

Comment (workingjubilee) [reply library/std/src/sys/sync/condvar/pthread.rs:28]: Then perhaps I need a remedial lesson on synchronization here, because the comments in OnceBox are very sparing and I'm not sure how the sync works out:
- https://github.com/rust-lang/rust/blob/c322cd5c5ac13b5ef16a487c889f8c148d8e3841/library/std/src/sys/sync/once_box.rs#L45-L73

**Comment by workingjubilee** [comment]
To put it another way:
- Last I checked, the standard library has internal functions with "this actually, literally has to only be called once in **the entire program**" as a safety constraint.
- This instead has a more loose sense of "once per mutex or condvar", and this is *fine*.
- However, even saying "once per mutex or condvar" is *ambiguous* because we have `sys::sync::Mutex`, `sys::pal::unix::sync::Mutex`, and `pthread_mutex_t`. There's also "the mutex" as in "the thing that wraps the OnceBox" and there's "the mutex as in the thing we allocate and then put behind the OnceBox's pointer". And likewise when considering condvars.

Thinking again, my understanding is that what happens here, for both types, can be, hypothetically:
1. Thread A acquire-loads and finds the OnceBox of a `sys::sync::Condvar` or `sys::sync::Mutex` is empty.
2. Thread A allocates `box_a`, putting the `pthread_mutex_t` or `pthread_condvar_t` in the allocated space.
3. Thread A initializes the `pthread_mutex_t` or `pthread_condvar_t` in the allocated space owned by `box_a`.
4. Thread B performs the same steps as 1-3, creating `box_b`[^1].
5. Finally, Thread A and Thread B both reach the AcqRel compare-exchange inside OnceBox, which forces them to sync. They play rock, paper, scissors over the memory bus for which gets to go first.
6. Today, Thread B wins, and writes `box_b`'s pointer to the OnceBox field (`.cvar` or `.pal`).
7. Thread A sees that Thread B won, and has to run the destructors of the data it has created. So it will tear down the data allocated in the space of `box_a` if necessary, and then deallocate the pointer to this space. This is fine because nothing further could possibly have happened to the `pthread_mutex_t` or `pthread_condvar_t` after initialization, as we haven't exposed its address to anywhere except the pthread library, so it's not in a state where this is a concern.

As I understand it, this is _sound_, but we have objectively *run the initialization function twice*, and thus the safety comment is not really the clearest. We can't even say "once per condvar" or "once per mutex", because of how we have three different things we are calling a "mutex" or "condvar", so it's ambiguous, and "the thing we can only initialize once" is *each place in memory*. The job of this implementation is only to present the *appearance* that we've only allocated and initialized one `pthread_mutex_t` or `pthread_condvar_t`, not the *reality* that we have done so. My understanding is this is acceptable because we're calling functions that care about the _allocated places in memory, behind the pointers_ we are operating on.

**This is all fine**. I just would like this to be clear from the comments, because we are three layers deep in terms of signifier vs. signified, we have indirections on top of indirections. The safety comments should be written more from the "insider perspective", since they're, well, inside the implementation.

[^1]: Note that as long as both hit the acquire-load before they hit the **next** step, the compare-exchange, then **neither are synchronized yet**.

File: library/std/src/sys/sync/condvar/pthread.rs (hunk 1 above)
Comment (workingjubilee) [reply library/std/src/sys/sync/condvar/pthread.rs:28]: I pried it apart.

I don't like "per object" as a constraint, because we have objects that manage objects, and we may create two objects to put inside one object, and then we have to get rid of one object, but we keep the other object, because we put it inside of the object...

... ... ...you get the picture.

What matters is that _the pointee_ only sees one init.

File: library/std/src/sys/pal/unix/sync/mutex.rs [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        Mutex { inner: UnsafeCell::new(libc::PTHREAD_MUTEX_INITIALIZER) }
+    }
+
+    pub(super) fn raw(&self) -> *mut libc::pthread_mutex_t {
+        self.inner.get()
+    }
+
+    /// # Safety
+    /// Must only be called once.
+    pub unsafe fn init(self: Pin<&mut Self>) {
```
Comment (workingjubilee) [inline library/std/src/sys/pal/unix/sync/mutex.rs:22]:
```suggestion
    /// This value must only be initialized once.
    pub unsafe fn init(self: Pin<&mut Self>) {
```

File: library/std/src/sys/pal/unix/sync/condvar.rs [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+    target_os = "l4re",
+    target_os = "redox",
+    target_os = "teeos",
+))]
+impl Condvar {
+    pub const PRECISE_TIMEOUT: bool = false;
+    const CLOCK: libc::clockid_t = libc::CLOCK_REALTIME;
+
+    /// # Safety
+    /// May only be called once.
```
Comment (workingjubilee) [inline library/std/src/sys/pal/unix/sync/condvar.rs:140]:
```suggestion
    /// # Safety
    /// This value must only be initialized once.
```

File: library/std/src/sys/pal/unix/sync/condvar.rs [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    target_os = "l4re",
+    target_os = "redox",
+    target_os = "teeos",
+)))]
+impl Condvar {
+    pub const PRECISE_TIMEOUT: bool = true;
+    const CLOCK: libc::clockid_t = libc::CLOCK_MONOTONIC;
+
+    /// # Safety
+    /// May only be called once.
```
Comment (workingjubilee) [inline library/std/src/sys/pal/unix/sync/condvar.rs:98]:
```suggestion
    /// This value must only be initialized once.
```

File: library/std/src/sys/sync/condvar/pthread.rs [hunk 7]
```diff
@@ -1,196 +1,88 @@
...
+        Condvar { cvar: OnceBox::new(), mutex: AtomicUsize::new(0) }
     }
 
-    fn get(&self) -> *mut libc::pthread_cond_t {
-        self.inner.get_or_init(AllocatedCondvar::new).0.get()
+    #[inline]
+    fn get(&self) -> Pin<&pal::Condvar> {
+        self.cvar.get_or_init(|| {
+            let mut cvar = Box::pin(pal::Condvar::new());
+            // SAFETY: we only call `init` once, namely here.
```
Comment (workingjubilee) [inline library/std/src/sys/sync/condvar/pthread.rs:25]:
```suggestion
            // SAFETY: pointee is initialized exactly once, by this closure
```

File: library/std/src/sys/sync/mutex/pthread.rs [hunk 8]
```diff
@@ -1,163 +1,66 @@
...
-    fn get(&self) -> *mut libc::pthread_mutex_t {
-        // If initialization fails, the mutex is destroyed. This is always sound,
-        // however, as the mutex cannot have been locked yet.
-        self.inner.get_or_init(AllocatedMutex::new).0.get()
+    fn get(&self) -> Pin<&pal::Mutex> {
+        // If the initialization race is lost, the new mutex is destroyed.
+        // This is sound however, as it cannot have been locked.
+        self.pal.get_or_init(|| {
+            let mut pal = Box::pin(pal::Mutex::new());
+            // SAFETY: we only call `init` once, namely here.
```
Comment (workingjubilee) [inline library/std/src/sys/sync/mutex/pthread.rs:24]:
```suggestion
            // SAFETY: pointee is initialized exactly once, by this closure
```

**Review by workingjubilee — COMMENTED** [review]
A possible diff that removes the ambiguity.

**Comment by joboet** [comment]
It's less about places in memory (you could safely destroy a `pal::Mutex` and initialize a new one in its place), but about instances of `pal::Mutex`/`pal::Condvar`. `init` may only be called once per instance of those. I've updated the comments to clarify this, hopefully that helps.

**Review by workingjubilee — APPROVED** [review]
thank you.
